hosting/src/global/Home/Home.ts

```diff
@@ -29,62 +29,98 @@ export default class Home extends Vue {
+      if (firstSlide === null) return
+
+      // Old => Exit to right
       activeSlide.classList.remove('translate-x-0')
       activeSlide.classList.add('translate-x-full')
-      nextSlide.classList.remove('-translate-x-full')
-      nextSlide.classList.add('translate-x-0')
+
+      // New => Enter from left
+      firstSlide.classList.remove('-translate-x-full')
+      firstSlide.classList.add('translate-x-0')
+
+      // Waiting => Ready to move on from right
       const otherSlides = document.querySelectorAll('.slide.-translate-x-full')
       otherSlides.forEach(element => {
         element.classList.remove('-translate-x-full')
         element.classList.add('translate-x-full')
       })
     } else {
+      // Old => Exit to left
       activeSlide.classList.remove('translate-x-0')
       activeSlide.classList.add('-translate-x-full')
+
+      // New => Enter from right
       nextSlide.classList.remove('translate-x-full')
       nextSlide.classList.add('translate-x-0')
     }
-
-    this.resetSlideTimer()
   }
 
-  // TODO: This should wrap-around left
   previousSlide(): void {
+    if (this.banners.length < 2) return
+
     const activeSlide = document.querySelector('.slide.translate-x-0')
-    const previousSlide = activeSlide?.previousElementSibling
-    if (activeSlide === null || previousSlide === null || previousSlide === undefined) return
-    activeSlide.classList.remove('translate-x-0')
-    activeSlide.classList.add('translate-x-full')
-    previousSlide.classList.remove('-translate-x-full')
-    previousSlide.classList.add('translate-x-0')
-
-    this.resetSlideTimer()
+    if (activeSlide === null) return // no slides
+
+    const previousSlide = activeSlide.previousElementSibling
+    if (previousSlide === null || !previousSlide.classList.contains('slide')) {
+      // Return to end...
+      const allSlides = activeSlide.parentElement?.querySelectorAll('.slide') ?? []
+      const lastSlide = allSlides.length > 0 ? allSlides[allSlides.length - 1] : null
+      if (lastSlide === null) return
+
+      // Old => Exit to left
+      activeSlide.classList.remove('translate-x-0')
+      activeSlide.classList.add('-translate-x-full')
+
+      // New => Enter from right
+      lastSlide.classList.remove('translate-x-full')
+      lastSlide.classList.add('translate-x-0')
+
+      // Waiting => Ready to move on from left
+      const otherSlides = document.querySelectorAll('.slide.translate-x-full')
+      otherSlides.forEach(element => {
+        element.classList.remove('translate-x-full')
+        element.classList.add('-translate-x-full')
+      })
+    } else {
+      // Old => Exit to right
+      activeSlide.classList.remove('translate-x-0')
+      activeSlide.classList.add('translate-x-full')
+
+      // New => Enter from left
+      previousSlide.classList.remove('-translate-x-full')
+      previousSlide.classList.add('translate-x-0')
+    }
```

Comment:
I think you can add `return` after if case (after for each loop). So, you don't need the last `else`. you can just attach

```
// Old => Exit to right
activeSlide.classList.remove('translate-x-0')
activeSlide.classList.add('translate-x-full')

// New => Enter from left
previousSlide.classList.remove('-translate-x-full')
previousSlide.classList.add('translate-x-0')
```
After that if case with `return`.

Reply:
I hadn't considered that. I inherited this structure from @antonyharfield. 

I think I prefer to limit early returns to guards. This if block is significantly longer. The else communicates clearly that the bottom 4 lines only execute on 1 branch.

That said... if we invert the if, we also simplify it:
```ts
if (previousSlide?.classList?.contains('slide')) ...
```

I might also consider creating methods like:
```ts
slideInRight(slide: Element): boolean {
  slide.classList.remove('-translate-x-full')
  slide.classList.add('translate-x-0')
  return true
}

slideOutRight(slide: Element): boolean {
  slide.classList.remove('translate-x-0')
  slide.classList.add('translate-x-full')
  return true
}
```

Then perhaps it could be more guard-like:
```ts
if (...) return slideOutLeft(activeSlide) && slideInLeft(previousSlide)
```

Reply:
> if (...) return slideOutLeft(activeSlide) && slideInLeft(previousSlide)
This also look cool!

I don't have many experience on the banner(I think it will look like carousal?). Normally, I do the logic by checking the index of the current banner, so it will slide forward and backward. I do that by create 2 function e.g. back() and next(), then if it is last (checking by array length) then just rolling back to first banner again like looping.

Reply:
It turns out the if is *really* ugly with our lint rules:
```ts
if(previousSlide && previousSlide. classList.contains('slide')) // lint complains 
if(previousSlide?. classList?.contains('slide') ?? false) // TS couldn't infer that slide isn't null
if(previousSlide?. classList?.contains('slide') === true) // ewww
```

Yes, it's a carousel.

Yes, I think holding an array of "slide" elements & the current index is easier. I'm going to refactor to that, using `ref`. I might also wrap the HTML elements in a class with methods like `slideToLeft()`, `slideToMiddle`, etc.

And I think it's finally time to extract the carousel to its own component.

Reply:
I think I should also add a new section to our PR template... "what is this feature? (background/introduction)" to make it easier for reviewers to understand the PR.

Reply:
Replaced this code; see [new discussion](https://github.com/phits-tech/website/pull/43/files#r628838693)